src/hotspot/share/services/heapDumperCompression.cpp

```diff
@@ -250,24 +250,17 @@ void CompressionBackend::deactivate() {
     ml.notify_all();
   }
 
-  // Wait for the threads to drain the compression work list.
+  // Wait for the threads to drain the compression work list and do some work yourself.
   while (!_to_compress.is_empty()) {
-    // If we have no threads, compress the current one itself.
-    if (_nr_of_threads == 0) {
-      MutexUnlocker mu(_lock, Mutex::_no_safepoint_check_flag);
-      thread_loop(true);
-    } else {
-      ml.wait();
-    }
+    do_foreground_work();
   }
 
   _active = false;
   ml.notify_all();
 }
 
-void CompressionBackend::thread_loop(bool single_run) {
-  // Register if this is a worker thread.
-  if (!single_run) {
+void CompressionBackend::thread_loop() {
```

Comment:
BTW: why is `ml.notify_all()` in line 275 needed at all?

Reply:
Hi,

thanks for the review Lin and Richard. 

The notify_all() is indeed not needed anymore. It was originally needed when the worker threads were newly created threads and we had to wait for them to finish at the end of the dump operation. But since we now use the GC work gang, this can be removed.

I will update the PR with your suggestions.

Best regards,
Ralf